lib/Sema/CSApply.cpp

```diff
@@ -3022,9 +3041,17 @@ namespace {
     }
 
     Expr *visitApplyExpr(ApplyExpr *expr) {
-      return finishApply(expr, cs.getType(expr),
-                         ConstraintLocatorBuilder(
-                           cs.getConstraintLocator(expr)));
+      auto *calleeLoc = CalleeLocators[expr];
+      assert(calleeLoc);
+
+      // Resolve the callee for the application if we have one.
+      ConcreteDeclRef callee;
+      if (auto overload = solution.getOverloadChoiceIfAvailable(calleeLoc)) {
+        auto *decl = overload->choice.getDeclOrNull();
+        callee = resolveConcreteDeclRef(decl, calleeLoc);
+      }
```

Comment:
I'm not entirely happy with the fact that we cache the resolved concrete decl refs on `ExprRewriter`, as it means we can't just use `resolveLocatorToDecl` here. I can't immediately think of a better alternative though. Ideally we would cache them on the `Solution`, but that would require a `const_cast` which I believe would be UB if we ever had a `const Solution`.

Reply:
I think that this is _okay_ but might be good to add a comment for posterity to outline the reason why we have so many `resolveConcreteDeclRef`s :)

Reply:
Thanks for the review! I've gone ahead and added a few comments, do they look good?

Reply:
Looks good, thank you!